src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java

```diff
@@ -0,0 +1,98 @@
+package com.es.phoneshop.model.product.DAO;
+
+import com.es.phoneshop.model.product.entity.Product;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class ArrayListProductDao implements ProductDao {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Lock readLock = lock.readLock();
+    private final Lock writeLock = lock.writeLock();
+    private List<Product> products;
+
+    public ArrayListProductDao() {
+        products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
+    @Override
+    public Product getProduct(Long id) {
+        readLock.lock();//�� ������� ��� �� ������ ����� ���, �� ��� ������� ������������ ������ �����
+        try {
+            return products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(NoSuchElementException::new);
+        } finally {
+            readLock.unlock();
+        }
+
+    }
+
+    @Override
+    public List<Product> findProducts() {
+        readLock.lock();
+        try {
+            return products.stream()
+                    .filter(product -> product.getPrice() != null)
+                    .filter(product -> product.getStock() > 0)
+                    .collect(Collectors.toList());
+        } finally {
+            readLock.unlock();
+        }
+
+    }
+
+    @Override
+    public Product save(Product product) {
+        writeLock.lock();
+        try {
+            if (product.getId() != null) {
+                products.set((int) (product.getId()-1), product);
+            } else {
+                product.setId((long) products.size() + 1);
```

Comment:
UUID for id generation
DAO must not store any manipulations on "Product" POJO, all initializations of Product must be defined in its own class, in constructor